Added CI integration + Docker.

AlexSkrypnyk opened the pull request:
## Changes
- Added Docker stack
- Added support for code linting + config
- Added support for unit tests (nosetests)
- Added support for BDD tests (behave)
- Updated documentation

## How to test this PR
- Checkout this branch
- Follow README (validate the steps there)
- Run all important `ahoy` commands: `build`, 'lint', 'test-unit`, `test-bdd`
- Test that site is accessible
- Test that changing files on the host machine is reflected witihin a container when reloading a page (e.g., change some text in the config form).
- Make sure that switches in `.env` file work. (e.g. set `ALLOW_BDD_FAIL=1` then run `ahoy test-bdd && echo $?` and check the exit code - 0 = good, 1 = bad).

AlexSkrypnyk committed:
Added CI integration + Docker.

MarkCalvert reviewed (approved):
All passed apart from ahoy test-unit with the below error
```
nose.plugins.cover: ERROR: Coverage not available: unable to import coverage module
Traceback (most recent call last):
  File "/app/ckan/default/bin/nosetests", line 10, in <module>
    sys.exit(run_exit())
  File "/app/ckan/default/lib/python2.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python2.7/unittest/main.py", line 94, in __init__
    self.parseArgs(argv)
  File "/app/ckan/default/lib/python2.7/site-packages/nose/core.py", line 145, in parseArgs
    self.config.configure(argv, doc=self.usage())
  File "/app/ckan/default/lib/python2.7/site-packages/nose/config.py", line 347, in configure
    self.plugins.begin()
  File "/app/ckan/default/lib/python2.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/app/ckan/default/lib/python2.7/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/app/ckan/default/lib/python2.7/site-packages/pylons/test.py", line 74, in begin
    relative_to=path)
  File "/app/ckan/default/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
    return loadobj(APP, uri, name=name, **kw)
  File "/app/ckan/default/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 271, in loadobj
    global_conf=global_conf)
  File "/app/ckan/default/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 296, in loadcontext
    global_conf=global_conf)
  File "/app/ckan/default/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 317, in _loadconfig
    loader = ConfigLoader(path)
  File "/app/ckan/default/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 393, in __init__
    with open(filename) as f:
IOError: [Errno 2] No such file or directory: '/app/test.ini'
```

AlexSkrypnyk committed:
Fixed tests.

AlexSkrypnyk committed:
t2

salsa-nathan committed:
[DQL-80] CI Tidy-ups (#3)

- Added extension dependencies for `ckanext-data-qld`, `ckanext-data-qld-theme` &  `ckanext-ytp-comments`
- Added dev dependencies for extensions to `requirements-dev.txt`
- Added necessary config settings to `test.ini` CKAN configuration file

Commit:
Added BDD tests and setup mailmock for email tests

Commit:
Added comment for mailmock debugging

Commit:
Updated wait time for mock emails

Commit:
Changed env config to allow unit test failures to run bdd tests

Commit:
Updated dev-requirements to install the expected nose_parameterized version the unit tests where written for

Commit:
Reverted back to orginal get_config_bool_value code to fix unit tests

AlexSkrypnyk commented on test/features/steps/steps.py:
```diff
@@ -0,0 +1,44 @@
+from behaving.mail.steps import *
+import random
+
+@when('I go to homepage')
```
`@step`

duttonw replied:
hard to say if it should be "when", or a "then" or a "and" per https://cucumber.io/docs/gherkin/reference/#steps

maybe a 'and' would be more suitable as it fit under a given. 
given persona 'x' logs in
and i go to the homepage
when the homepage has finished loading
then i should see that the logout button
but i should not see the login button

AlexSkrypnyk commented on test/features/steps/steps.py:
```diff
@@ -0,0 +1,44 @@
+    when_i_visit_url(context, '/')
+
+
+@when('I log in')
```
`@step`

AlexSkrypnyk commented on test/features/environment.py:
```diff
@@ -0,0 +1,106 @@
+        email=u'ckan_user@localhost',
+        password=u'password'
+    ),
+    'SalsaAdmin': dict(
```
please replace all references to `Salsa` with `Test`

duttonw replied:
Is SalsaAdmin/SalsaEditor an org admin?
If so change it to UserWithOrgAdminGroup or simliar for its persona

AlexSkrypnyk commented on test/features/datarequest.feature:
```diff
@@ -0,0 +1,102 @@
+@smoke
```
remove `@smoke`

AlexSkrypnyk commented on test/features/datarequest.feature:
```diff
@@ -0,0 +1,102 @@
+@smoke
+Feature: Datarequest
+
+# DQL-26: Data Requests - Install and configure extensionUAT - the Data Requests are accessible via the /datarequest URL (does not display a 404 error)
```
here and below - please remove all references to internal ticket numbers - it does not make sense to keep internal references in open source project. Also, these comments become stale very fast.

As for the scenario title - please use whatever is in the comment as a Scenario title.

duttonw replied:
Scenario titles are should be as long as possible and describe exactly the behaviour its trying to accomplish.
i.e.
# DQL-27: Data Base & Form updatesUAT - when visiting the /datarequests page as a non-logged in user, the button at the top of the page reads "Login to create a data request" and links to the /user/login page
    Scenario: View 'Login to create data request' button when not logged in

should be
Scenario: Given an anonymous user visits, when they visit a datarequest page, Then a button is displayed at the top of the page with the text "Login to create a data request" and when clicked redirects the user to the Login Page (/user/login)

AlexSkrypnyk reviewed (commented):
Please see my comment.

Also, please reinstate homepage and login tests - they are used to test theBDD framework itself.

duttonw reviewed (approved):
cleanup does need to happen but overall its a good step forward in the testing front.

Commit:
Adding back in homepage and login features
Updated testing data

Commit:
Added ckan background worker to run for email tests

Commit:
Added new email tests for data requests
Updated bdd tests to decode base64 email messages

Commit:
Fixed stoping CKAN background job worker

duttonw commented on test/features/datarequest.feature:
```diff
@@ -0,0 +1,150 @@
+Feature: Datarequest
+
+# Data Requests - Install and configure extensionUAT - the Data Requests are accessible via the /datarequest URL (does not display a 404 error)
+    Scenario: Browse to data request page
```
comments needs to be added to scenario name. This needs to be done for all of them. 
Just add a - and append to the end.

i.e. 
Browse to data request page
 becomes
Browse to data request page - Data Requests - Install and configure extensionUAT - the Data Requests are accessible via the /datarequest URL (does not display a 404 error)

Please do to all of them

duttonw commented on test/features/login.feature:
```diff
@@ -0,0 +1,20 @@
+@smoke	
+Feature: Login	
+
+     Scenario: Login process	
```
scenario names needs more information on what its actually doing (i.e. login with sysadmin i expect to log out button and take screenshot)

duttonw commented on test/features/login.feature:
```diff
@@ -0,0 +1,20 @@
+        And I take a screenshot	
+        Then I should see an element with xpath "//a[contains(string(), 'Log out')]"	
+
+     Scenario: Login step	
```
is this login failure?  or is this testing the login function?

duttonw commented on test/features/steps/steps.py:
```diff
@@ -0,0 +1,59 @@
+
+@when('I go to homepage')
+def go_to_home(context):
+    when_i_visit_url(context, '/')
```
when_i_visit_url 

should be renamed to go_to_page

duttonw commented on test/features/steps/steps.py:
```diff
@@ -0,0 +1,59 @@
+# Modifed the default step to decode the payload from base64
+@then(u'I should receive an email at "{address}" containing "{text}"')
+def should_receive_email_containing_text(context, address, text):
+    def filter_contents(mail):
```
maybe useful to not have inside this function, and wrapped in a if just in case the contents is of other states. (maybe use a python library for email reading? as there is at least 5 type of content encoding)

https://www.w3.org/Protocols/rfc1341/5_Content-Transfer-Encoding.html
Content-Transfer-Encoding := "BASE64" / "QUOTED-PRINTABLE" / "8BIT"   / "7BIT" / "BINARY" / x-token

content-transfer-encoding: quoted-printable 
looks like this
```
<!DOCTYPE html>=0D
<html>=0D
<head><meta name=3D"viewport" content=3D"width=3Ddevice-width"><meta http-e=
quiv=3D"Content-Type" content=3D"text/html; charset=3DUTF-8">=0D
etc
```

duttonw reviewed (changes requested):
overall, very good work. Some minor improvements and it would be great.

Also for the screenshots, i may have missed this but is travis' ci (or other) collecting these into a zip for later download/review.

Would still like to see a bit more in the testing front where screenshots only occur by a flag instead of all the time or if error occurs.

Commit:
Changed when steps to steps.
Updated email test step name

Commit:
Updated CKAN version to 2.8.3

Commit:
Updated scenario names to be more descriptive
Added scenario tags

Commit:
Fixing password change used by tests

salsa-nathan committed:
Update plugin.py

salsa-nathan committed:
Update plugin.py
